core/shared/src/main/scala/zio/ZIOFunctionConstructor.scala

```diff
@@ -0,0 +1,104 @@
+package zio
+
+/**
+ * A `ZIOFunctionConstructor[Input]` knows how to construct a `ZIO` value from a
+ * function of type `In`. This allows the type of the `ZIO` value constructed to
+ * depend on `In`.
+ */
+sealed abstract class ZIOFunctionConstructor[In] {
+
+  /**
+   * The type of the `ZIO` value.
+   */
+  type Out
+
+  /**
+   * Constructs a `ZIO` value from the specified input.
+   */
+  def apply(in: In)(implicit trace: Trace): Out
+}
+
+object ZIOFunctionConstructor {
+
+  type WithOut[In, Out0] = ZIOFunctionConstructor[In] { type Out = Out0 }
+
+  implicit def function1Constructor[A: Tag, R, E, Z: Tag]
```

Comment:
I think you missed my previous comment. The `: Tag` on `Z` is unnecessary as far as I can tell (in this and other methods)

Reply:
missed it. done